### PR TITLE
fixes #3138 - removing rbvmomi require statement

### PR DESCRIPTION
**Pull request by lzap**

### CODE REVIEW

**Comment by lzap** [comment]
Pls ignore this PR as I am not done yet with this, will push proper implementation soon.

**Comment by lzap** [comment]
Okay now it works fine. Please test against vmware instance, if you have one handy.

**Comment by ohadlevy** [comment]
is there a reason why we should not do this to every CR? i think its the same for ovirt (with rbovirt and GCE

**Comment by lzap** [comment]
If there is not, I will do that.

**Comment by domcleal** [comment]
I agree with @ohadlevy, we've talked about it before and this would be great on every compute resource type if the code's made generic.  I thought we had a redmine ticket for it, but can't find it now.  Most CR types depend on the fog group at least, others on fog + another group.

I'd expected this to be solved like https://github.com/lzap/foreman/blob/48ecae3a7dc02e8882b15fc942b4647500477654/app/services/foreman/provision/ssh.rb#L1 (sorry, I gave you the wrong example earlier), but this is a much better way, so :+1: if you want to complete it this way.

**Comment by lzap** [comment]
AS SEEN ON TV:

about

**Comment by lzap** [comment]
Note previously smart proxies and compute resources tabs were rendered only if there was at least one SP and one CR. I changed that behavior - if there is a SP new tab appears, if there is a CR new tab appears. AP is always there.

Yeah, next step would be plugin registry... :-)

**Comment by ohadlevy** [comment]
ping @abenari ?

**Comment by domcleal** [comment]
Pretty!

File: bundler.d/gce.rb [hunk 1]
```diff
@@ -1,4 +1,4 @@
 group :gce do
-  gem "google-api-client"
+  gem "google-api-client", :require => "google/api_client"
```
Comment (lzap) [inline bundler.d/gce.rb:2]: FYI This was needed for my bundler, GCE was not automatically loaded. And we will need that for bundler_ext support anyway.

File: app/models/compute_resource.rb [hunk 2]
```diff
@@ -3,7 +3,13 @@ class ComputeResource < ActiveRecord::Base
   include Taxonomix
   include Encryptable
   encrypts :password
-  PROVIDERS = %w[ Libvirt Ovirt EC2 Vmware Openstack Rackspace GCE].delete_if{|p| p == "Libvirt" && !SETTINGS[:libvirt]}
+  SUPPORTED_PROVIDERS = %w[Libvirt Ovirt EC2 Vmware Openstack Rackspace GCE]
```
Comment (domcleal) [inline app/models/compute_resource.rb:6]: [done] All of these also depend on the fog group being loaded, so we should check for that too.

File: app/models/compute_resource.rb [hunk 3]
```diff
@@ -3,7 +3,13 @@ class ComputeResource < ActiveRecord::Base
...
+  PROVIDERS = SUPPORTED_PROVIDERS.dup; PROVIDERS.delete_if do |p|
+    (p == "Libvirt" && !SETTINGS[:libvirt]) ||
```
Comment (domcleal) [inline app/models/compute_resource.rb]: Can we just check for `SETTINGS[p.downcase.to_sym]` instead of repeating?

File: app/views/about/index.html.erb [hunk 4]
```diff
@@ -1,17 +1,41 @@
...
+            <table class="table table-striped">
+              <tr>
+                <th><%= _("Provider") %></th>
+                <th><%= _("Status") %></th>
+              </tr>
+              <% ComputeResource::SUPPORTED_PROVIDERS.each do |provider| %>
+                <tr>
+                  <td><%= provider %></td>
+                  <% if ComputeResource::PROVIDERS.include?(provider) %>
+                    <td><div class="badge badge-success">Installed</div></td>
```
Comment (domcleal) [inline app/views/about/index.html.erb]: extract strings here

File: app/views/about/index.html.erb [hunk 5]
```diff
@@ -1,17 +1,41 @@
...
+                  <% else %>
+                    <td><div class="badge badge-inverse">Not Installed</div></td>
+                  <% end %>
+                </tr>
+              <% end %>
+            </table>
+            <%= _('To enable a provider either install OS package (e.g. foreman-libvirt) or enable bundler group for development setup (e.g. :ovirt).') %>
+          </div>
+          <% if @proxies.any? %>
+          <div class="tab-pane" id="smart_proxies">
```
Comment (domcleal) [inline app/views/about/index.html.erb]: This should be indented as it's within a conditional.

File: app/views/about/index.html.erb [hunk 6]
```diff
@@ -1,17 +1,41 @@
 <% title _("About") %>
 
 <div class="row" id="about">
-  <% if @proxies.any? || @compute_resources.any? %>
     <div class="span7">
```
Comment (domcleal) [inline app/views/about/index.html.erb]: Unindent?

File: config/application.rb [hunk 7]
```diff
@@ -18,16 +18,23 @@
...
-    begin
-      Bundler.require(:libvirt) if SETTINGS[:unattended]
-    rescue LoadError
-      puts "Libvirt bindings are missing - hypervisor management is disabled"
+    if SETTINGS[:unattended]
+      %w[libvirt ovirt vmware gce].each do |group|
+        begin
+          Bundler.require(group)
+        rescue LoadError
+          puts "Bindings for #{group} are missing - compute resource is disabled"
```
Comment (domcleal) [inline config/application.rb]: I'd really like to kill this message, or use the logger if we can at this stage?  It's noisy as it is (with rake tasks etc), but this will be far worse.

File: config/application.rb [hunk 8]
```diff
@@ -18,16 +18,23 @@
...
+
+      SETTINGS[:libvirt] = defined?(::Libvirt) && true
+      SETTINGS[:ovirt] = defined?(::OVIRT) && true
+      SETTINGS[:vmware] = defined?(::RbVmomi) && true
+      SETTINGS[:gce] = defined?(::Google::APIClient::VERSION) && true
     end
   end
 end
 
-SETTINGS[:libvirt] = SETTINGS[:unattended] && defined?(Libvirt)
```
Comment (domcleal) [inline config/application.rb:29]: [fixed] I think it was deliberately here to work with bundler_ext, since the code path above that they're now in won't be run under bundler_ext, so the SETTINGS never get set.

File: config/application.rb [hunk 9]
```diff
@@ -18,16 +18,23 @@
...
+    if SETTINGS[:unattended]
+      %w[libvirt ovirt vmware gce].each do |group|
+        begin
+          Bundler.require(group)
+        rescue LoadError
+          puts "Bindings for #{group} are missing - compute resource is disabled"
+        end
+      end
+
+      SETTINGS[:libvirt] = defined?(::Libvirt) && true
```
Comment (domcleal) [inline config/application.rb]: Why `&& true`?

File: config/application.rb [hunk 10]
```diff
@@ -18,16 +18,23 @@
       console {Foreman.setup_console}
     end
     Bundler.require(*Rails.groups(:assets => %w(development test)))
-    begin
-      Bundler.require(:libvirt) if SETTINGS[:unattended]
-    rescue LoadError
-      puts "Libvirt bindings are missing - hypervisor management is disabled"
+    if SETTINGS[:unattended]
+      %w[libvirt ovirt vmware gce].each do |group|
```
Comment (domcleal) [inline config/application.rb]: Let's add fog to this.

File: app/models/compute_resource.rb (hunk 2 above)
Comment (lzap) [inline app/models/compute_resource.rb:6]: Check what? Check if ::Fog constant is available?

File: app/models/compute_resource.rb (hunk 3 above)
Comment (lzap) [inline app/models/compute_resource.rb]: Yeah I also changed that to use regexp which is nicer.

File: app/views/about/index.html.erb (hunk 4 above)
Comment (lzap) [inline app/views/about/index.html.erb]: !

File: app/views/about/index.html.erb (hunk 6 above)
Comment (lzap) [inline app/views/about/index.html.erb]: Heh, yeah the whole file deserves indentitation love. Let me do that in a separate commit once you are all happy with the patch. Do not forget to remind me before merging :-)

File: config/application.rb (hunk 9 above)
Comment (lzap) [inline config/application.rb]: Cos otherwise it reads "constant"?

File: app/models/compute_resource.rb (hunk 2 above)
Comment (domcleal) [inline app/models/compute_resource.rb:6]: Yes, whether say libvirt works depends on having both Fog and Libvirt and whether EC2 works depends just on Fog.

File: config/application.rb (hunk 9 above)
Comment (domcleal) [inline config/application.rb]: Ah okay, same as `!!` then.

File: config/application.rb (hunk 7 above)
Comment (lzap) [inline config/application.rb]: It seems logger is not ready (does not print anything at this stage), so removing.

Pushed all the changes.

File: app/models/compute_resource.rb [hunk 11]
```diff
@@ -3,7 +3,10 @@ class ComputeResource < ActiveRecord::Base
   include Taxonomix
   include Encryptable
   encrypts :password
-  PROVIDERS = %w[ Libvirt Ovirt EC2 Vmware Openstack Rackspace GCE].delete_if{|p| p == "Libvirt" && !SETTINGS[:libvirt]}
+  SUPPORTED_PROVIDERS = %w[Libvirt Ovirt EC2 Vmware Openstack Rackspace GCE]
+  PROVIDERS = SUPPORTED_PROVIDERS.dup; PROVIDERS.delete_if do |p|
```
Comment (domcleal) [inline app/models/compute_resource.rb]: `PROVIDERS = SUPPORTED_PROVIDERS.reject do ...` will do the same, without modifying SUPPORTED_PROVIDERS

File: app/models/compute_resource.rb [hunk 12]
```diff
@@ -3,7 +3,10 @@ class ComputeResource < ActiveRecord::Base
...
+    p =~ /^(Libvirt|Ovirt|GCE|Vmware)$/ && !SETTINGS[p.downcase.to_sym]
```
Comment (domcleal) [inline app/models/compute_resource.rb]: `%w[Libvirt Ovirt Vmware GCE].include?(p)` rather than regex?

**Comment by domcleal** [comment]
So yep, this looks good - but having the CRs appear conditionally based on the presence of Fog should be part of it I think.

**Comment by lzap** [comment]
Right, all fixed now. Ty.

## LZ

Later,

 Lukas "lzap" Zapletal
 irc: lzap #theforeman

**Comment by domcleal** [comment]
@lzap I don't think this handles the CRs that only depend on Fog, just those that depend on Fog _and_ some other library?

**Comment by lzap** [comment]
Well that block is now awful... :-D

## 

Later,

 Lukas "lzap" Zapletal
 irc: lzap #theforeman

File: app/models/compute_resource.rb [hunk 13]
```diff
@@ -3,7 +3,10 @@ class ComputeResource < ActiveRecord::Base
   include Taxonomix
   include Encryptable
   encrypts :password
-  PROVIDERS = %w[ Libvirt Ovirt EC2 Vmware Openstack Rackspace GCE].delete_if{|p| p == "Libvirt" && !SETTINGS[:libvirt]}
+  SUPPORTED_PROVIDERS = %w[Libvirt Ovirt EC2 Vmware Openstack Rackspace GCE]
+  PROVIDERS = SUPPORTED_PROVIDERS.reject do |p|
+    %w[Libvirt Ovirt Vmware GCE].include?(p) && !SETTINGS[p.downcase.to_sym]
```
Comment (domcleal) [inline app/models/compute_resource.rb]: I think we can remove this first conditional now.

File: config/application.rb [hunk 14]
```diff
@@ -18,15 +18,23 @@
...
     end
   end
 end
 
-SETTINGS[:libvirt] = SETTINGS[:unattended] && defined?(Libvirt)
+SETTINGS[:libvirt] = defined?(::Fog) && defined?(::Libvirt)
+SETTINGS[:ovirt] = defined?(::Fog) && defined?(::OVIRT)
+SETTINGS[:vmware] = defined?(::Fog) && defined?(::RbVmomi)
+SETTINGS[:gce] = !! defined?(::Google::APIClient::VERSION)
+SETTINGS[:openstack] = SETTINGS[:rackspace] = SETTINGS[:gce] = SETTINGS[:ec2] = !! defined?(::Fog)
```
Comment (domcleal) [inline config/application.rb]: gce should be removed from here

File: config/application.rb [hunk 15]
```diff
@@ -18,15 +18,23 @@
...
+      end
     end
   end
 end
 
-SETTINGS[:libvirt] = SETTINGS[:unattended] && defined?(Libvirt)
+SETTINGS[:libvirt] = defined?(::Fog) && defined?(::Libvirt)
+SETTINGS[:ovirt] = defined?(::Fog) && defined?(::OVIRT)
+SETTINGS[:vmware] = defined?(::Fog) && defined?(::RbVmomi)
+SETTINGS[:gce] = !! defined?(::Google::APIClient::VERSION)
```
Comment (domcleal) [inline config/application.rb]: Missing `defined?(::Fog)`

**Comment by domcleal** [comment]
Some minor bits above, but it tests ok so please go ahead with the indentation fixes.

**Comment by lzap** [comment]
Done :-)

Jesus Christ, and I was only going to remove one requirement here :-P

## 

Later,

 Lukas "lzap" Zapletal
 irc: lzap #theforeman

**Comment by domcleal** [comment]
@lzap that's why I was surprised (happy-surprised) that you wanted to do it this way originally :)

lzap++

**Comment by domcleal** [comment]
Thanks @lzap, merged as d31ea09 for Foreman 1.4.0.
